src/saas_shield/standard.rs

```diff
@@ -277,6 +404,59 @@ impl StandardDocumentOps for SaasShieldStandardClient {
             edek: plaintext_document.edek,
         })
     }
+
+    /// Encrypt multiple documents with the provided metadata.
+    /// The provided EDEKs will be decrypted and used to encrypt each corresponding document's fields.
+    /// This is useful when updating some fields of the document.
+    async fn encrypt_with_existing_edek_batch(
+        &self,
+        plaintext_documents: PlaintextDocumentsWithEdeks,
+        metadata: &AlloyMetadata,
+    ) -> Result<StandardEncryptBatchResult, AlloyError> {
+        let edeks_with_headers = plaintext_documents
+            .iter()
+            .map(|(k, v)| (k.as_str(), v.edek.clone()));
+        let (mut batch_unwrap_response, edek_failures) = self
+            .batch_unwrap_edeks(edeks_with_headers, metadata)
+            .await?;
+        let docs_and_keys = plaintext_documents
+            .into_iter()
+            .map(|(document_id, document)| {
+                let maybe_keys = batch_unwrap_response.keys.remove(&document_id);
+                (document_id, (document, maybe_keys))
+            });
+        let encrypt_document = |(plaintext_document, maybe_keys): (
+            PlaintextDocumentWithEdek,
+            Option<UnwrapKeyResponse>,
+        )| {
+            let UnwrapKeyResponse { dek } = maybe_keys.ok_or_else(|| AlloyError::DecryptError {
+                msg: "TSP failed to wrap key for document.".to_string(),
+            })?;
+            let edek_parts = Self::decompose_edek_header(plaintext_document.edek)?;
+            let enc_key = tsc_dek_to_encryption_key(dek.0)?;
+            edek_parts.validate_signature(enc_key)?;
+            Self::encrypt_document(
+                self.rng.clone(),
+                edek_parts.get_edek_bytes()?,
+                enc_key,
+                &metadata.tenant_id,
+                plaintext_document.document,
+            )
+        };
+        let decryption_result = collection_to_batch_result(docs_and_keys, encrypt_document);
+        let combined_failures = batch_unwrap_response
+            .failures
+            .into_iter()
+            .map(|(k, err)| (k, err.into()))
+            .chain(edek_failures)
+            .chain(decryption_result.failures)
+            .collect();
```

Comment:
It seems like we've done this same pattern for merging failures several times now. Could we make a trait that allows it or a util function `merge_batch_failures` or something?

Reply:
I looked into this and couldn't find a nice way to do it. The pattern is used in 3 places and is slightly different among each, so I think I'd rather leave it for now